internal/handlers/pg/pjson/pjson.go

```diff
@@ -151,9 +172,79 @@ func toPJSON(v any) pjsontype {
 	panic(fmt.Sprintf("not reached: %T", v)) // for go-sumtype to work
 }
 
-// Unmarshal decodes the given pjson-encoded data.
-func Unmarshal(data []byte) (any, error) {
-	var v any
+// Unmarshal decodes the top-level document.
+// It decodes document's schema from the `$s` field and uses it to decode the data of the document.
+func Unmarshal(data []byte) (*types.Document, error) {
+	var v map[string]json.RawMessage
+	r := bytes.NewReader(data)
+	dec := json.NewDecoder(r)
+
+	err := dec.Decode(&v)
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	if err = checkConsumed(dec, r); err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	// decode schema from the $s field of the document
+	jsch, ok := v["$s"]
+	if !ok {
+		return nil, lazyerrors.Errorf("schema is not set")
+	}
+
+	var sch schema
+	r = bytes.NewReader(jsch)
+	dec = json.NewDecoder(r)
+	dec.DisallowUnknownFields()
+
+	if err := dec.Decode(&sch); err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	if err := checkConsumed(dec, r); err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	delete(v, "$s")
+
+	// decode data from the rest of the document using the schema
+	if len(sch.Keys) != len(v) {
+		return nil, lazyerrors.Errorf("pjson.Unmarshal: document must have the same number of keys and values (keys: %d, values: %d)", len(sch.Keys), len(v))
```

Comment:
Not exactly this, but we can be specific about schema keys and document keys mismatch :hugs:
```suggestion
		return nil, lazyerrors.Errorf("pjson.Unmarshal: document must have the same number of keys as schema (schema: %d, keys: %d)", len(sch.Keys), len(v))
```

Reply:
fixed